File: pkg/src/bezier.c
/* $ID: bezier.c, last updated 2024-09-03, F.Osorio */

#include "fastmatrix.h"

void
bezier_smoother(double *x, double *y, int *n, double *grid, int *ngrid, double *xgrid, double *ygrid)
{ /* computation of Bezier smoother */
  double s, *work;

  if (*ngrid <= 0) /* quick return if possible */
    return;

  work = (double *) R_Calloc(2, double);
  for (int i = 0; i < *ngrid; i++) {
    s = grid[i];
    F77_CALL(decasteljau)(x, y, n, &s, work);
    xgrid[i] = work[0];
    ygrid[i] = work[1];
  }
  R_Free(work);
}

File: pkg/R/bezier.R
## ID: bezier.R, last updated 2021-11-14, F.Osorio

bezier <- function(x, y, ngrid = 200)
{ ## smoothing via Bezier curve

  # validating arguments
  if (!is.vector(x))
    x <- as.vector(x)
  if (!is.numeric(x))
    stop("argument x is not a numeric vector")
  if (!is.vector(y))
    y <- as.vector(y)
  if (!is.numeric(y))
    stop("argument y is not a numeric vector")

  # remove all NAs
  OK <- complete.cases(x, y)
  x <- x[OK]
  y <- y[OK]

  n <- length(x)
  if (n != length(y))
    stop("arguments x and y do not have the same length.")

  grid <- seq(0, 1, length = ngrid)

  z <- .C("bezier_smoother",
          x = as.double(x),
          y = as.double(y),
          n = as.integer(n),
          grid = as.double(grid),
          ngrid = as.integer(ngrid),
          xgrid = double(ngrid),
          ygrid = double(ngrid))[c("xgrid","ygrid")]
  z
}
